kotest-property/src/commonMain/kotlin/io/kotest/property/config.kt

```diff
@@ -25,13 +26,37 @@ fun PropTest.toPropTestConfig() =
       minSuccess = minSuccess,
       maxFailure = maxFailure,
       iterations = iterations,
-      shrinkingMode = shrinkingMode
+      shrinkingMode = shrinkingMode,
+      listeners = listeners
    )
 
 data class PropTestConfig(
    val seed: Long? = null,
    val minSuccess: Int = Int.MAX_VALUE,
    val maxFailure: Int = 0,
    val shrinkingMode: ShrinkingMode = ShrinkingMode.Bounded(1000),
-   val iterations: Int? = null
-)
+   val iterations: Int? = null,
+   val listeners: MutableList<PropTestListener> = mutableListOf()
+) {
+   fun beforeTest(before: suspend () -> Unit) {
+      listeners.add(object : PropTestListener {
+         override suspend fun beforeTest() {
+            before()
+            super.beforeTest()
+         }
+      })
+   }
+   fun afterTest(after: suspend () -> Unit) {
```

Comment:
I don't like the idea of having this class have mutable state.
Kotest is built along FP principles so we should keep this class immutable to avoid surprises for users. 
The only mutable classes we have are either internal and not visible to users, or in the project config god object.
I know that means the nice DSL methods can't exist here but that's a trade off that we must accept.

Reply:
Ok. If you like having the list in there (and not in a different object), I had another solution with List (and listOf). Let me know and I’ll refactor.